packages/ui/src/components/Visualization/Canvas/Canvas.tsx

```diff
@@ -22,17 +22,12 @@ interface CanvasProps {
 
 export const Canvas: FunctionComponent<PropsWithChildren<CanvasProps>> = (props) => {
   const [selectedIds, setSelectedIds] = useState<string[]>([]);
+  const [controller, setController] = useState<Visualization>(CanvasService.createController());
 
-  const controller = useMemo(() => {
-    const newController = CanvasService.createController();
-    newController.addEventListener(SELECTION_EVENT, setSelectedIds);
-    newController.addEventListener(GRAPH_LAYOUT_END_EVENT, () =>
-      /** TODO: Schedule the layouting in a better fashion */
-      setTimeout(() => {
-        newController.getGraph().fit(80);
-      }, 100),
-    );
-    return newController;
+  // Set up the controller one time
+  useEffect(() => {
+    controller.addEventListener(SELECTION_EVENT, setSelectedIds);
+    setController(controller);
   }, []);
```

Comment:
I'm happy that we could remove that sketchy `setTimeout` 👏 

I'm just noticing now that we were not returning a cleanup function for the `addEventListener` calls. I was doing some tests and I think we could use something like the following:

```suggestion
  const controller = useMemo(() => CanvasService.createController(), []);

  // Set up the controller one time
  useEffect(() => {
    const localController = controller;
    const graphLayoutEndFn = () => {
      localController.getGraph().fit(80);
    };

    localController.addEventListener(SELECTION_EVENT, setSelectedIds);
    localController.addEventListener(GRAPH_LAYOUT_END_EVENT, graphLayoutEndFn);

    return () => {
      localController.removeEventListener(SELECTION_EVENT, setSelectedIds);
      localController.removeEventListener(GRAPH_LAYOUT_END_EVENT, graphLayoutEndFn);
    };
  }, []);
```

Let me go through the code.

1. Here we create the controller once, but instead of using `useState`, we're using `useMemo` since we won't need to update this later.
```typescript
  const controller = useMemo(() => CanvasService.createController(), []);
```

2. Here we capture the reference to that memoed controller as this will come in handy for the cleanup function and later on, we define the callback that will be called when the rendering finishes. We do the latter to keep a reference for the cleanup function later.
```typescript
    const localController = controller;
    const graphLayoutEndFn = () => {
      localController.getGraph().fit(80);
    };
```

3. Finally we set the event listeners and return a cleanup function for the `useEffect` callback. Notice how within this function we call `.removeEventListener()` using the referenced controller and the functions, this way, if the reference changes, we will be pointing to the right one.
```typescript
    localController.addEventListener(SELECTION_EVENT, setSelectedIds);
    localController.addEventListener(GRAPH_LAYOUT_END_EVENT, graphLayoutEndFn);

    return () => {
      localController.removeEventListener(SELECTION_EVENT, setSelectedIds);
      localController.removeEventListener(GRAPH_LAYOUT_END_EVENT, graphLayoutEndFn);
    };
```

Reply:
Thanks for the suggestion, updated the PR according to your suggestion.